stl/inc/flat_set

```diff
@@ -424,63 +447,95 @@ private:
     }
 
     template <class _Ty>
-        requires (_Keylt_transparent && is_constructible_v<_Kty, _Ty>) || is_same_v<_Ty, _Kty>
-    void _Emplace_hint(const_iterator _Where, _Ty&& _Val) {
-        _Container& _Cont     = _Get_cont();
-        _Keylt& _Compare      = _Get_comp();
-        const iterator _Begin = begin();
-        const iterator _End   = end();
+    auto _Emplace(_Ty&& _Val) {
+        _Container& _Cont = _Get_cont();
+        if constexpr (_Multi) {
+            _STL_INTERNAL_STATIC_ASSERT(is_same_v<remove_cvref_t<_Ty>, _Kty>);
+            return _Cont.emplace(upper_bound(_Val), _STD forward<_Ty>(_Val));
+        } else {
+            const iterator _End   = end();
+            const iterator _Where = lower_bound(_Val);
+            if (_Where != _End && _Keys_equal(*_Where, _Val)) {
+                return pair{_Where, false};
+            }
+            if constexpr (is_same_v<remove_cvref_t<_Ty>, _Kty>) {
+                return pair{_Cont.emplace(_Where, _STD forward<_Ty>(_Val)), true};
+            } else {
+                // flat_set::insert(auto&&)
+                _STL_INTERNAL_STATIC_ASSERT(_Keylt_transparent && is_constructible_v<_Kty, _Ty>);
+                _Kty _Keyval{_STD forward<_Ty>(_Val)};
+                _STL_ASSERT(lower_bound(_Keyval) == _Where && !_Keys_equal(_Keyval, *_Where),
+                    "find(val) was not equal to find(key_type{forward(val)})");

```

Comment:
This assertion is to avoid invariant breakage. It can capture errorneous usages like:
```cpp
std::flat_set<int, std::less<>> s{ 1,2,3,5,7 };
s.insert(2.5);
```
`Which is UB as find(2.5)->end, but find(int(2.5))->2`; However, I'm highly uncertain what this precondition implies to the implementation:
![image](https://github.com/microsoft/STL/assets/60953653/fd62e7d3-6761-45ed-8b9d-26d94f73535c)
Is `flat_set::insert([hint,]auto&&)` really well-defined by the standard?

Reply:
For example, can I assume that `lower_bound(val)==lower_bound(keyval)`? If so, how is this guaranteed? if not, then how should `hint` be used?

Reply:
Also, for the first overload `insert(K&& x)`, the `bool` cannot indicate whether `x` is moved if it was rvalue, as it's unclear whether the `emplace` must success.

Reply:
My latest understanding is that, `less<>` is not supposed to be used as comparators for `flat_meow`, as it's too general. The acceptable types should be limited to a set of pre-defined types, and should be compared as if comparing the same thing (for example, there are `foo::key`, `bar::key` and `baz::key`, and the transparent comparator should compare them as if by comparing the `key`.) So `lower_bound(val)==lower_bound(key{forward(val)})` should really hold true, and `emplace` is supposed to succeed after finding no equivalent keys.